sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,347 @@
+
+namespace Microsoft.Azure.ContainerRegistry
+{
+    /*Robust handling of Basic and OAUTH2 authentication flows for the Azure Container Registry Runtime .Net SDK.
+     This class handles Basic Authentication as well as JWT token authentication using both username and password
+     routes as well as through exchanging AAD tokens. */
+    public class AcrClientCredentials : ServiceClientCredentials
+    {
+
+        #region Definitions
+        private class TokenCredentials : ServiceClientCredentials
+        {
+            private string AuthHeader { get; set; }
+
+            /*To be used for General Login Scheme*/
+            public TokenCredentials(string username, string password)
+            {
+                AuthHeader = EncodeTo64(username + ":" + password);
+            }
+            /*To be used for exchanging AAD Tokens for ACR Tokens*/
+            public TokenCredentials()
+            {
+                AuthHeader = null;
+            }
+            public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                if (request == null)
+                {
+                    throw new ArgumentNullException("request");
+                }
+                if (AuthHeader != null)
+                {
+                    request.Headers.Authorization = new AuthenticationHeaderValue("Basic", AuthHeader);
+                }
+                await base.ProcessHttpRequestAsync(request, cancellationToken);
+            }
+        }
+
+        public struct Token
+        {
+            public string TokenStr { get; set; }
+            public DateTime Expiration { get; set; }
+        }
+
+        private static readonly JwtSecurityTokenHandler JwtSecurityClient = new JwtSecurityTokenHandler();
+
+        public static Token MakeToken(string token)
+        {
+            Token tok = new Token();
+            tok.TokenStr = token;
+            SecurityToken fields = JwtSecurityClient.ReadToken(token);
+            tok.Expiration = fields.ValidTo;
+            return tok;
+        }
+
+        public enum LoginMode
+        {
+            Basic,
+            TokenAuth,
+            TokenAad
+        }
+
+        // Constant to refresh tokens slighly before they are to expire guarding against possible latency related crashes
+        public const double LATENCY_SAFETY = 2;
+
+        public delegate Token AADAcquireCallback();
+        #endregion
+
+        #region Instance Variables        
+        private string AuthHeader { get; set; }
+        private LoginMode Mode { get; set; }
+        private string LoginUrl { get; set; }
+        private string Username { get; set; }
+        private string Password { get; set; }
+        private String Tenant { get; set; }
+        private CancellationToken RequestCancellationToken { get; set; }
+
+        // Structure : Scope : Token
+        private Dictionary<string, Token> AcrAccessTokens;
```

Comment:
>AcrAccessTokens [](start = 42, length = 15)

camelCase all private fields if the SDK has a coding guidelines. Also can you elide the commonInit method and just initialize them here?